File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/Direction.java
package pize.tests.voxelgame.main;

import pize.math.vecmath.vector.Vec2i;
import pize.math.vecmath.vector.Vec3i;

public enum Direction{
    
    NEGATIVE_X(-1,  0,  0),
    POSITIVE_X( 1,  0,  0),
    NEGATIVE_Y( 0, -1,  0),
    POSITIVE_Y( 0,  1,  0),
    NEGATIVE_Z( 0,  0, -1),
    POSITIVE_Z( 0,  0,  1),
    NONE      ( 0,  0,  0);
    
    
    private final Vec3i normal;
    
    Direction(int x, int y, int z){
        normal = new Vec3i(x, y, z);
    }

    public final Vec3i getNormal(){
        return normal;
    }


    public static Direction fromNormal(Vec3i normal){
        return fromNormal(normal.x, normal.y, normal.z);
    }

    public static Direction fromNormal(int x, int y, int z){
        if(x == 1)
            return POSITIVE_X;
        else if(y == 1)
            return POSITIVE_Y;
        else if(z == 1)
            return POSITIVE_Z;
        else if(x == -1)
            return NEGATIVE_X;
        else if(y == -1)
            return NEGATIVE_Y;
        else if(z == -1)
            return NEGATIVE_Z;
        else
            return NONE;
    }


    private static final Vec2i[] normal2D = new Vec2i[]{
        new Vec2i(-1,  0), new Vec2i( 1,  0), new Vec2i( 0, -1), new Vec2i( 0,  1)
    };

    private static final Vec3i[] normal3D = new Vec3i[]{
        NEGATIVE_X.getNormal(), POSITIVE_X.getNormal(), NEGATIVE_Y.getNormal(), POSITIVE_Y.getNormal(), NEGATIVE_Z.getNormal(), POSITIVE_Z.getNormal()
    };

    public static Vec2i normal2DFromIndex(int index){
        return normal2D[index];
    }

    public static Vec3i normal3DFromIndex(int index){
        return normal3D[index];
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/Direction.kt
package pize.tests.voxelgame.main

import pize.math.vecmath.vector.Vec2i
import pize.math.vecmath.vector.Vec3i

enum class Direction(x: Int, y: Int, z: Int) {
    NEGATIVE_X(-1, 0, 0),
    POSITIVE_X(1, 0, 0),
    NEGATIVE_Y(0, -1, 0),
    POSITIVE_Y(0, 1, 0),
    NEGATIVE_Z(0, 0, -1),
    POSITIVE_Z(0, 0, 1),
    NONE(0, 0, 0);

    val normal: Vec3i

    init {
        normal = Vec3i(x, y, z)
    }

    companion object {
        fun fromNormal(normal: Vec3i): Direction {
            return fromNormal(normal.x, normal.y, normal.z)
        }

        fun fromNormal(x: Int, y: Int, z: Int): Direction {
            return if (x == 1) POSITIVE_X else if (y == 1) POSITIVE_Y else if (z == 1) POSITIVE_Z else if (x == -1) NEGATIVE_X else if (y == -1) NEGATIVE_Y else if (z == -1) NEGATIVE_Z else NONE
        }

        private val normal2D = arrayOf(
            Vec2i(-1, 0), Vec2i(1, 0), Vec2i(0, -1), Vec2i(0, 1)
        )
        private val normal3D = arrayOf(
            NEGATIVE_X.normal,
            POSITIVE_X.normal,
            NEGATIVE_Y.normal,
            POSITIVE_Y.normal,
            NEGATIVE_Z.normal,
            POSITIVE_Z.normal
        )

        fun normal2DFromIndex(index: Int): Vec2i {
            return normal2D[index]
        }

        fun normal3DFromIndex(index: Int): Vec3i {
            return normal3D[index]
        }
    }
}
